### DISCUSSION TITLE
One-button scrolling through spaces [Moved to Issue #352]

### DISCUSSION THREAD

**Discussion by sjdennis** [noah-nuebling/mac-mouse-fix#138] [Other]
There is one really annoying limitation with scrolling through spaces in Mac OS - you have to go left, or right. You can't just have a single button that toggles through all of them, which would be far more logical - I never remember whether a space is left or right of the one I am in now - and would free up one of those buttons for another function. This is a Mac OS problem, not a mac-mouse-fix problem - they've made their own interface annoying for no reason. 

But I wonder if there is a workaround. Is there any way of the software finding out when you have reached the end of the available spaces, so it could then automatically switch to scrolling in the other direction? Or might there be a hidden system call that actually does scroll through spaces like this but isn't made clear to the user, which the sofware could access? 

Probably there's nothing that can be done, but if someone can think of a way to emulate something like this it would be awesome.

**Comment by MuhammedZakir** [comment]
Have you tried HammerSpoon?

1. (deprecated api) https://www.hammerspoon.org/docs/hs.spaces.watcher.html
2. (undocumented api) https://github.com/asmagill/hs._asm.undocumented.spaces

**Comment by noah-nuebling** [comment]
Implementing a feature like this should be possible using the private SkyLight framework. 

This is the best attempt I’ve found at reverse engineering the Skylight framework:
https://github.com/NUIKit/CGSInternal.

Unfortunately it’s far from complete and might be out of date in certain places. Also, the currently discovered APIs could break in future macOS updates. But MMF and many other apps depend on this and it has been fine for many years.

The „(undocumented API)“ hammerspoon module which @MuhammedZakir mentioned is essentially a wrapper around this as far as I understand.

---

Edit: I just checked and unfortunately, the APIs so far discovered in the NUIKit repo don’t seem to provide an obvious way to see if you're on the first or last space. Maybe we could figure out a way by playing around with the existing APIs, but from looking into them superficially, I think that's unlikely.

However, I’m sure that there’s a way using the SkyLight framework. We just need to discover more of its API. If anyones interested in that, you can read more here https://github.com/NUIKit/CGSInternal/issues/2

**Reply by MuhammedZakir** [reply]
In asmagill's `hs._asm.undocumented.spaces` module, there are functions to get the current space's ID and also for getting a list of spaces in arranged order. Using these work, but may break in a future update.

```lua
-- Returns the ID of the currently active space.
spaces.activeSpace()
```

```lua
-- Returns a table whose keys are the screenUUID of the available screens.
-- Each key's value is an array of the spaces on that display,
-- in the order in which they are currently arranged.
spaces.layout()
```

---

@noah-nuebling : if it works (it does<sup>[1]</sup>), will you implement this? You probably have to add a note that this is an expermintal feature  and may break in a future macOS update.

[1] it may not work well with multiple monitors.

**Comment by noah-nuebling** [comment]
Hi! I moved this topic to Issue #352. Please continue any discussion there. Thanks!
